### ISSUE TITLE
c4dbobs_getChanges entries getting invalidated too early

### ISSUE THREAD

**Issue by borrrden** [couchbase/couchbase-lite-core#418]
Observed on Windows, Mac, and Linux via .NET:  Threading CreateVsEnumerate bindings test fails with an error indicating that the docID it received in a `C4DatabaseChange` is invalid.  It then prints out a bunch of garbage as the contents of said change.  The commit that caused this is 0c9cc20c074290c61c8b0615c2bb735a5a7449e7 and I've tracked it down to one specific line:  

```c++
void SequenceTracker::addExternalTransaction(const SequenceTracker &other) {
        Assert(!inTransaction());
        Assert(other.inTransaction());
        if (!_changes.empty() || _numDocObservers > 0) {
            log("addExternalTransaction from %s", other.loggingIdentifier().c_str());
            for (auto e = next(other._transaction->_placeholder); e != other._changes.end(); ++e) {
                _lastSequence = e->sequence;
                _documentChanged(e->docID, e->revID, e->sequence, e->bodySize);
            }
            removeObsoleteEntries(); // <-- This line causes the bad behavior.  Without it the test passes.
        }
    }
```

Is the indicated line needed there?  If not can we remove it?

**Comment by borrrden** [comment]
I'm going to comment out this line because I need this bad behavior gone.  If this is a problem please reopen this ticket with the reason why.

**Comment by snej** [comment]
That `removeObsoleteEntries()` call will prevent the changes list from growing without bounds in some circumstances. It shouldn't have caused any problems … can you give any more details about what circumstance causes the problem?

**Comment by snej** [comment]
I understand the problem now. The fix above takes care of the memory issue but reintroduces the leak that I fixed with 0c9cc20. We need a better fix...

>[5:17 PM] Jens Alfke: I think I see what's going wrong, now.
>readChanges is returning the right data ... but the memory backing those slices (docID and revID) can be freed by removeObsoleteEntries.
>The problem happens when
>- main thread calls readChanges
>- background thread commits a transaction
>- bg thread calls addExternalTransaction on main thread's tracker
>- this calls removeObsoleteEntries, which frees some of the docID slices
>- now the main thread examines the docIDs it got from readChanges X_X
